unlock.go

```diff
@@ -115,52 +131,147 @@ type ReferentialUnlock interface {
 	seenSigUnlocks := map[uint16]struct{}{}
+	seenSigBlockPubkeyBytes := map[string]int{}
+	seenSigBlockPubkeyBytesInMultiUnlocks := map[string]int{}
 	seenRefUnlocks := map[uint16]ReferentialUnlock{}
-	seenSigUnlockBytes := map[string]int{}
+	seenMultiUnlocks := map[uint16]struct{}{}
+	seenMultiUnlockBytes := map[string]int{}
 
-	return func(index int, unlock Unlock) error {
-		switch x := unlock.(type) {
+	return func(index int, u Unlock) error {
+		switch unlock := u.(type) {
 		case *SignatureUnlock:
-			if x.Signature == nil {
+			if unlock.Signature == nil {
 				return ierrors.Wrapf(ErrSigUnlockHasNilSig, "at index %d is nil", index)
 			}
 
-			sigBlockBytes, err := api.Encode(x.Signature)
+			sigBlockPubKeyBytes, err := publicKeyBytesFromSignatureBlock(unlock.Signature)
 			if err != nil {
-				return ierrors.Errorf("unable to serialize signature unlock block at index %d for dup check: %w", index, err)
+				return ierrors.Wrapf(err, "unable to parse pubkey bytes from signature unlock block at index %d for dup check", index)
 			}
 
-			if existingIndex, exists := seenSigUnlockBytes[string(sigBlockBytes)]; exists {
+			// we check for duplicated pubkeys in SignatureUnlock(s)
+			if existingIndex, exists := seenSigBlockPubkeyBytes[string(sigBlockPubKeyBytes)]; exists {
 				return ierrors.Wrapf(ErrSigUnlockNotUnique, "signature unlock block at index %d is the same as %d", index, existingIndex)
 			}
 
-			seenSigUnlockBytes[string(sigBlockBytes)] = index
+			// we also need to check for duplicated pubkeys in MultiUnlock(s)
+			if existingIndex, exists := seenSigBlockPubkeyBytesInMultiUnlocks[string(sigBlockPubKeyBytes)]; exists {
+				return ierrors.Wrapf(ErrSigUnlockNotUnique, "signature unlock block at index %d is the same as in multi unlock at index %d", index, existingIndex)
+			}
+
 			seenSigUnlocks[uint16(index)] = struct{}{}
+			seenSigBlockPubkeyBytes[string(sigBlockPubKeyBytes)] = index
+
 		case ReferentialUnlock:
-			if prevRef := seenRefUnlocks[x.Ref()]; prevRef != nil {
-				if !x.Chainable() {
-					return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references existing referential unlock %d but it does not support chaining", index, x.Ref())
+			if prevRef := seenRefUnlocks[unlock.Ref()]; prevRef != nil {
+				if !unlock.Chainable() {
+					return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references existing referential unlock %d but it does not support chaining", index, unlock.Ref())
 				}
-				seenRefUnlocks[uint16(index)] = x
+				seenRefUnlocks[uint16(index)] = unlock
 
 				break
 			}
-			// must reference a sig unlock here
-			if _, has := seenSigUnlocks[x.Ref()]; !has {
-				return ierrors.Wrapf(ErrReferentialUnlockInvalid, "%d references non existent unlock %d", index, x.Ref())
+
+			// must reference a sig or multi unlock here
+			_, hasSigUnlock := seenSigUnlocks[unlock.Ref()]
+			_, hasMultiUnlock := seenMultiUnlocks[unlock.Ref()]
+			if !hasSigUnlock && !hasMultiUnlock {
```

Comment:
```suggestion
			_, isSigUnlock := seenSigUnlocks[unlock.Ref()]
			_, isMultiUnlock := seenMultiUnlocks[unlock.Ref()]
			if !isSigUnlock && !isMultiUnlock {
```

Nit: I feel like "is" would be more readable than "has" here.

Reply:
isSigUnlock sounds more like a typecast. I would keep it as it is.